cgen: passing an array into `(array).contains()`

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 12.3.1, 21E258
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.1.6 (clang-1316.0.21.2)

getwd: /Users/hunam/Documents/dev/go2v
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-04-19 10:49:51
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 1fc54a1.a318a2e

Git version: git version 2.32.0 (Apple Git-132)
Git vroot status: weekly.2022.15-77-ga318a2e0
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
```

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
arr := [0]
[0].contains(arr)
```

**What did you expect to see?**
I don't know, maybe it should work, maybe it should throw an error
 
**What did you see instead?**
```v
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/a.6992593488331177315.tmp.c:11369:83: error: passing 'Array_int' (aka 'struct array') to parameter of incompatible type 'int'
        Array_int_contains(new_array_from_c_array(1, 1, sizeof(int), _MOV((int[1]){0})), arr);
                                                                                         ^~~
/tmp/v_501/a.6992593488331177315.tmp.c:1923:49: note: passing argument to parameter 'v' here
static bool Array_int_contains(Array_int a, int v) {
                                                ^
1 warning and 1 error generated.
...
==================
```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1893,6 +1893,17 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 		node.return_type = ast.void_type
 	} else if method_name == 'contains' {
 		// c.warn('use `value in arr` instead of `arr.contains(value)`', node.pos)
+		if node.args.len != 1 {
+			c.error('`.contains()` expected 1 argument, but got $node.args.len', node.pos)
+		} else {
+			arg_typ := ast.mktyp(c.expr(node.args[0].expr))
+			elem_typ_str := c.table.type_to_str(elem_typ)
+			arg_typ_str := c.table.type_to_str(arg_typ)
+			if !left_sym.has_method('contains') && elem_typ_str != arg_typ_str {
+				c.error('`.contains()` expected `$elem_typ_str` argument, but got `$arg_typ_str`',
+					node.pos)
+			}
+		}
 		node.return_type = ast.bool_type
 	} else if method_name == 'index' {
 		node.return_type = ast.int_type

--- a/vlib/v/checker/tests/array_contains_args_err.out
+++ b/vlib/v/checker/tests/array_contains_args_err.out
@@ -1,0 +1,35 @@
+vlib/v/checker/tests/array_contains_args_err.vv:3:17: error: `.contains()` expected `int` argument, but got `[]int`
+    1 | fn main() {
+    2 |     arr := [0]
+    3 |     mut ret := [0].contains([0])
+      |                    ~~~~~~~~~~~~~
+    4 |     ret = [0].contains()
+    5 |     ret = [0, 1, 2].contains(0, 1, 2)
+vlib/v/checker/tests/array_contains_args_err.vv:4:12: error: `.contains()` expected 1 argument, but got 0
+    2 |     arr := [0]
+    3 |     mut ret := [0].contains([0])
+    4 |     ret = [0].contains()
+      |               ~~~~~~~~~~
+    5 |     ret = [0, 1, 2].contains(0, 1, 2)
+    6 |     ret = [0].contains('a')
+vlib/v/checker/tests/array_contains_args_err.vv:5:18: error: `.contains()` expected 1 argument, but got 3
+    3 |     mut ret := [0].contains([0])
+    4 |     ret = [0].contains()
+    5 |     ret = [0, 1, 2].contains(0, 1, 2)
+      |                     ~~~~~~~~~~~~~~~~~
+    6 |     ret = [0].contains('a')
+    7 |     ret = [0].contains(arr)
+vlib/v/checker/tests/array_contains_args_err.vv:6:12: error: `.contains()` expected `int` argument, but got `string`
+    4 |     ret = [0].contains()
+    5 |     ret = [0, 1, 2].contains(0, 1, 2)
+    6 |     ret = [0].contains('a')
+      |               ~~~~~~~~~~~~~
+    7 |     ret = [0].contains(arr)
+    8 |     println(ret)
+vlib/v/checker/tests/array_contains_args_err.vv:7:12: error: `.contains()` expected `int` argument, but got `[]int`
+    5 |     ret = [0, 1, 2].contains(0, 1, 2)
+    6 |     ret = [0].contains('a')
+    7 |     ret = [0].contains(arr)
+      |               ~~~~~~~~~~~~~
+    8 |     println(ret)
+    9 | }

--- a/vlib/v/checker/tests/array_contains_args_err.vv
+++ b/vlib/v/checker/tests/array_contains_args_err.vv
@@ -1,0 +1,9 @@
+fn main() {
+	arr := [0]
+	mut ret := [0].contains([0])
+	ret = [0].contains()
+	ret = [0, 1, 2].contains(0, 1, 2)
+	ret = [0].contains('a')
+	ret = [0].contains(arr)
+	println(ret)
+}

